connectors/sources/graphql.py

```diff
@@ -0,0 +1,501 @@
+
+        if not self.validate_endpoints():
+            msg = "HTTP URL & Endpoint are not structured."
+            raise ConfigurableFieldValueError(msg)
+
+        if not self.is_query(graphql_query=self.graphql_client.graphql_query):
+            msg = "Configured Query is not supported by the connector."
+            raise ConfigurableFieldValueError(msg)
+
+        headers = self.graphql_client.configuration["headers"]
+        graphql_variables = self.graphql_client.configuration["graphql_variables"]
+        if headers:
+            try:
+                self.graphql_client.headers = json.loads(headers)
+            except Exception as exception:
+                msg = f"Error while processing configured GraphQL headers. Exception: {exception}"
+                raise ConfigurableFieldValueError(msg) from exception
+
+        if graphql_variables:
+            try:
+                self.graphql_client.variables = json.loads(graphql_variables)
+            except Exception as exception:
+                msg = f"Error while processing configured GraphQL variables. Exception: {exception}"
+                raise ConfigurableFieldValueError(msg) from exception
+
+    async def close(self):
+        await self.graphql_client.close()
+
+    async def ping(self):
+        try:
+            await self.graphql_client.ping()
+            self._logger.debug("Successfully connected to GraphQL Instance.")
+        except Exception:
+            self._logger.exception("Error while connecting to GraphQL Instance.")
+            raise
+
+    def yield_dict(self, documents):
+        if isinstance(documents, dict):
+            yield documents
+        elif isinstance(documents, list):
+            for document in documents:
+                if isinstance(document, dict):
+                    yield document
+
+    async def fetch_data(self, graphql_query):
+        if self.graphql_client.pagination_technique == "no_pagination":
+            data = await self.graphql_client.make_request(graphql_query=graphql_query)
+            for documents in self.graphql_client.extract_graphql_data_items(data=data):
+                for document in self.yield_dict(documents):
+                    yield document
+        else:
+            async for data in self.graphql_client.paginated_call(
+                graphql_query=graphql_query
+            ):
+                for document in self.yield_dict(data):
+                    yield document
+
+    async def get_docs(self, filtering=None):
+        """Executes the logic to fetch GraphQL response in async manner.
+
+        Args:
+            filtering (filtering, None): Filtering Rules. Defaults to None.
+
+        Yields:
+            dict: Documents from GraphQL.
+        """
+        doc_id = 1
+        async for doc in self.fetch_data(
+            graphql_query=self.graphql_client.graphql_query
+        ):
+            doc = deepcopy(doc)
+            doc["_id"] = doc.get("id") or hash_id(str(doc_id))
```

Comment:
I'm still not a fan of the `doc_id` approach. Given that we know exactly which objects to index (we have a path to find the key) can we somehow try to represent id as: `{path}-{hash of the dict}`?

Reply:
I think to convert whole dict into hash is not a better approach as some documents are too large. So, as dicussed on slack i guess we can move forward with above apporach for now and if i find anything helpful i will raise follow up PR.

Reply:
Given that we won't support incremental syncs I think it's acceptable to keep it that way for now - I can't think of good strategy given that graphQL doesn't have unified way of handling global IDs

Reply:
Actually let me check with the team how to handle that case, I will get back to you with decision.